### PR TITLE
Refactor Makefile

### PR DESCRIPTION
**Pull request by norio-nomura**
refactored and added targets:
- `guestagents` builds `lima-guestagent` for archs enabled by `CONFIG_GUESTAGENT_ARCHS_*`
- `native-guestagent` builds `lima-guestagent` for native arch
- `additional-guestagents` builds `lima-guestagent` for archs other than native arch
- `<arch>-guestagent` builds `lima-guestagent` for <arch>: aarch64 armv7l riscv64 x86_64
- `limactl` builds `limactl`, `codesign`, and `lima`
- `lima` copy `lima`, and `lima$(bat)`
- `helpers` copy `nerdctl.lima`, `apptainer.lima`, `docker.lima`, `podman.lima`, `kubectl.lima`
- `default_template` copy `examples/default.yaml`
- `templates` copy `examples/*.yaml`
- `template_experimentals` copy `examples/experimental/*.yaml`
- `create-examples-link` create link at, or copy to `_output/share/lima/examples`
- `documents` copy `*.md`, `LICENSE`, `SECURITY.md`, and `VERSION`
- `create-links-in-documents` create link at, or copy to `_output/share/doc/lima/{templates,examples}`

With this change, running `make clean limactl native-guestagent templates install` will allow you to install `limactl`, the native `linux-guestagent`, and all templates.

This is my solution to https://github.com/lima-vm/lima/issues/2426#issuecomment-2187316192

EDIT: added `<arch>-guestagent` targets

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -54,16 +54,30 @@ all: binaries manpages
...
+	@echo  '  templates                 - Copy templates'
+	@echo  '  template_experimentals    - Copy experimental templates'
+	@echo  '  default_template          - Copy default.yaml template'
+	@echo  '  create-examples-link      - Create a symlink at _output/share/lima/examples pointing to _output/share/lima/templates'
+	@echo
+	@echo  '  documents                 - Copy documents'
+	@echo  '  create-links-in-documents - Create symlinks in documents'
+	@echo
+	@echo  '  e.g. to install limactl, native guestagent, and templates:'
+	@echo  '   make clean limactl native-guestagent templates install'
```
Comment (jandubois) [inline Makefile:75]: Can we have a custom target for this, so I can just run `make local`? 😄

Comment (norio-nomura) [reply Makefile:75]: added `native` target

File: Makefile [hunk 2]
```diff
@@ -54,16 +54,31 @@ all: binaries manpages
...
+	@echo
+	@echo  '  limactl         - Build limactl, and lima'
+	@echo  '  lima            - Copy lima, and lima.bat'
+	@echo  '  helpers         - Copy nerdctl.lima, apptainer.lima, docker.lima, podman.lima, kubectl.lima'
+	@echo
+	@echo  '  guestagents               - Build guestagents for archs enabled by CONFIG_GUESTAGENT_ARCHS_*'
+	@echo  '  native-guestagent         - Build guestagent for native arch'
+	@echo  '  additional-guestagents    - Build guestagents for archs other than native arch'
+	@echo
+	@echo  '  templates                 - Copy templates'
```
Comment (AkihiroSuda) [inline Makefile:66]: Copy to where, for what purpose?

File: Makefile [hunk 3]
```diff
@@ -54,16 +54,31 @@ all: binaries manpages
 help:
```
Comment (AkihiroSuda) [inline Makefile:54]: Now the help is too long and unreadable

**Review by AkihiroSuda — COMMENTED** [review]
Why not just use plain cp?

File: Makefile (hunk 3 above)
Comment (norio-nomura) [reply Makefile:54]: I'll add `help-targets` target for added sub targets.

**Comment by norio-nomura** [comment]
> Why not just use plain cp?

By having `make` recognize each file, it can copy only those that have been modified.
e.g.
```console
➜ ~09:57:10 09:57:15~ (refactor-makefile) make templates
mkdir -p _output/share/lima/templates
cp -aL examples/README.md _output/share/lima/templates/README.md
cp -aL examples/almalinux-8.yaml _output/share/lima/templates/almalinux-8.yaml
cp -aL examples/almalinux-9.yaml _output/share/lima/templates/almalinux-9.yaml
cp -aL examples/almalinux.yaml _output/share/lima/templates/almalinux.yaml
cp -aL examples/alpine-image.yaml _output/share/lima/templates/alpine-image.yaml
cp -aL examples/alpine.yaml _output/share/lima/templates/alpine.yaml
cp -aL examples/apptainer-rootful.yaml _output/share/lima/templates/apptainer-rootful.yaml
cp -aL examples/apptainer.yaml _output/share/lima/templates/apptainer.yaml
cp -aL examples/archlinux.yaml _output/share/lima/templates/archlinux.yaml
cp -aL examples/buildkit.yaml _output/share/lima/templates/buildkit.yaml
cp -aL examples/centos-stream-9.yaml _output/share/lima/templates/centos-stream-9.yaml
cp -aL examples/centos-stream.yaml _output/share/lima/templates/centos-stream.yaml
cp -aL examples/debian-11.yaml _output/share/lima/templates/debian-11.yaml
cp -aL examples/debian-12.yaml _output/share/lima/templates/debian-12.yaml
cp -aL examples/debian.yaml _output/share/lima/templates/debian.yaml
cp -aL examples/default.yaml _output/share/lima/templates/default.yaml
cp -aL examples/docker-rootful.yaml _output/share/lima/templates/docker-rootful.yaml
cp -aL examples/docker.yaml _output/share/lima/templates/docker.yaml
cp -aL examples/faasd.yaml _output/share/lima/templates/faasd.yaml
cp -aL examples/fedora.yaml _output/share/lima/templates/fedora.yaml
cp -aL examples/k3s.yaml _output/share/lima/templates/k3s.yaml
cp -aL examples/k8s.yaml _output/share/lima/templates/k8s.yaml
cp -aL examples/opensuse-leap.yaml _output/share/lima/templates/opensuse-leap.yaml
cp -aL examples/opensuse.yaml _output/share/lima/templates/opensuse.yaml
cp -aL examples/oraclelinux-8.yaml _output/share/lima/templates/oraclelinux-8.yaml
cp -aL examples/oraclelinux-9.yaml _output/share/lima/templates/oraclelinux-9.yaml
cp -aL examples/oraclelinux.yaml _output/share/lima/templates/oraclelinux.yaml
cp -aL examples/podman-rootful.yaml _output/share/lima/templates/podman-rootful.yaml
cp -aL examples/podman.yaml _output/share/lima/templates/podman.yaml
cp -aL examples/rocky-8.yaml _output/share/lima/templates/rocky-8.yaml
cp -aL examples/rocky-9.yaml _output/share/lima/templates/rocky-9.yaml
cp -aL examples/rocky.yaml _output/share/lima/templates/rocky.yaml
cp -aL examples/ubuntu-lts.yaml _output/share/lima/templates/ubuntu-lts.yaml
cp -aL examples/ubuntu.yaml _output/share/lima/templates/ubuntu.yaml
cp -aL examples/vmnet.yaml _output/share/lima/templates/vmnet.yaml
➜ ~09:57:15 09:57:18~ (refactor-makefile) limactl edit templates/docker.yaml --memory 8
➜ ~09:57:18 09:57:20~ (refactor-makefile) make templates
cp -aL examples/docker.yaml _output/share/lima/templates/docker.yaml
➜ ~09:57:20 09:57:20~ (refactor-makefile)
```

File: Makefile (hunk 3 above)
Comment (norio-nomura) [reply Makefile:54]: added `help-targets` target and reverted `help` to simpler as previous.

**Review by AkihiroSuda — APPROVED** [review]

**Comment by norio-nomura** [comment]
Sorry, I added `<arch>-guestagent` targets.
e.g.
```console
➜ ~10:56:10 13:20:12~ (refactor-makefile) make armv7l-guestagent
GOOS=linux GOARCH=arm GOARM=7 CGO_ENABLED=0 go build -ldflags="-s -w -X github.com/lima-vm/lima/pkg/version.Version=v1.0.0-alpha.0-95-g1efc42e1.m" -tags "" -o _output/share/lima/lima-guestagent.Linux-armv7l ./cmd/lima-guestagent
chmod 644 _output/share/lima/lima-guestagent.Linux-armv7l
```

**Comment by norio-nomura** [comment]
The `artifacts-*` targets rely on the `binaries` target, which is affected by `.config`. Given the nature of these targets, I’m wondering if they should be rewritten to avoid using the `binaries` target, so they won’t be influenced by `.config`. 🤔

**Review by afbjorklund — COMMENTED** [review]
Just a naming feedback. But if you don't plan on using Kconfig / config.mk, then maybe it should be removed?

File: Makefile [hunk 4]
```diff
@@ -79,122 +105,137 @@ menuconfig: Kconfig
...
-	CGO_ENABLED=1 $(GO_BUILD) -o $@ ./cmd/limactl
+# _output/share/doc/lima
+DOCUMENTS=$(addprefix _output/share/doc/lima/,$(wildcard *.md) LICENSE SECURITY.md VERSION)
 
-.PHONY: _output/share/lima/lima-guestagent.Linux-x86_64
-_output/share/lima/lima-guestagent.Linux-x86_64:
-	GOOS=linux GOARCH=amd64 CGO_ENABLED=0 $(GO_BUILD) -o $@ ./cmd/lima-guestagent
-	chmod 644 $@
+.PHONY: documents
+documents: $(DOCUMENTS)
```
Comment (afbjorklund) [inline Makefile:207]: The "documents" target should probably be called "documentation" ?

File: Makefile (hunk 3 above)
Comment (afbjorklund) [reply Makefile:54]: Main reason for "make help" was only to make "make menuconfig" happy

```

*** End of the configuration.
*** Execute 'make' to start the build or try 'make help'.
```

File: Makefile (hunk 4 above)
Comment (norio-nomura) [reply Makefile]: Thanks!
Changed to `documentation`.
Also changed `create-links-in-documents` to `create-links-in-doc-dir`.

**Comment by norio-nomura** [comment]
> But if you don't plan on using Kconfig / config.mk, then maybe it should be removed?

`.config` remains valid. I also wrote code for the `defconfig` target and for generating it without external shell scripts, but I haven’t included it in this PR. I'll leave the patch here.
```patch
diff --git forkSrcPrefix/Makefile forkDstPrefix/Makefile
index 675d434629507a5ed4ad8debf086ca344de34a8b..55be2b781e3b55b1a2c3b540a71c70b5a49d6e14 100644
--- forkSrcPrefix/Makefile
+++ forkDstPrefix/Makefile
@@ -98,6 +98,21 @@ menuconfig: Kconfig
 	MENUCONFIG_STYLE=aquatic \
 	$(KCONFIG_MCONF) $<
 
+# Generate config for the current architecture (only)
+# This is done to avoid a dependency on KCONFIG tools
+defconfig: config_guestagent_arch = $(call config_guestagent_arch_name,$(1))=$(2)
+defconfig: CONFIG_GUESTAGENT_ARCHS=$(call config_guestagent_arch,$(NATIVE_GUESTAGENT_ARCH),y)
+defconfig: CONFIG_GUESTAGENT_ARCHS+=$(foreach arch,$(ADDITIONAL_GUESTAGENT_ARCHS),$(call config_guestagent_arch,$(arch),n))
+# Helper to join words with separator
+# usage: $(call join_sep,$(separator),$(words))
+defconfig: join_sep = $(subst $(space),$(1),$(2))
+defconfig: space = $(empty) $(empty)
+defconfig: empty =
+defconfig:
+	echo "CONFIG_GUESTAGENT_OS_LINUX=y" > .config
+	echo "$(call join_sep,\n,$(sort $(CONFIG_GUESTAGENT_ARCHS)))" >> .config
+	echo "CONFIG_GUESTAGENT_COMPRESS=n" >> .config
+
 # Copy the default config, if not overridden locally
 # This is done to avoid a dependency on KCONFIG tools
 .config: config.mk
 ```

File: Makefile (hunk 4 above)
Comment (norio-nomura) [reply Makefile]: Also changed `DOCUMENTS` variable name to `DOCUMENTATION`.

**Comment by norio-nomura** [comment]
rebased on master.

**Review by jandubois — COMMENTED** [review]
I think this PR is good now, but needs commits squashed.

@afbjorklund @AkihiroSuda Do you still have any concerns that need to be addressed, or can this be merged now?

**Comment by norio-nomura** [comment]
> , but needs commits squashed.

Instead of requiring PRs to be squashed before merging, wouldn't it be better to switch to using [squash merging](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests)? If the PR branch itself is squashed, the individual commits won't be retained anywhere unless you specifically ensure they are. With squash merging, the commits will still be accessible later via something like `refs/pull/2638/head` on GitHub, so you can reference them if needed.

By adding the following configuration to Git, the PR branch will become accessible.
```ini
[remote "upstream"]
	url = https://github.com/lima-vm/lima.git
	fetch = +refs/heads/*:refs/remotes/upstream/*
	# PR branches
	fetch = +refs/pull/*/head:refs/remotes/upstream/pull/*
	# PR branches' merge commit
	# fetch = +refs/pull/*/merge:refs/remotes/upstream/pull/merge/*
	gh-resolved = base
```

**Comment by jandubois** [comment]
> Instead of requiring PRs to be squashed before merging, wouldn't it be better to switch to using [squash merging](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests)?

It is fine with me, but I don't know if this is acceptable to @AkihiroSuda.

**Comment by AkihiroSuda** [comment]
> With squash merging, the commits will still be accessible later via something like refs/pull/2638/head on GitHub, so you can reference them if needed.

IIUC this is untrue when the PR branch is removed?

**Comment by norio-nomura** [comment]
> IIUC this is untrue when the PR branch is removed?

You can't view it via the GitHub web interface, but you can access it with Git.  
I deleted my `norio-nomura/vz-support-kernel-image` branch, but `upstream/refs/heads/pull/2562` is still available for checkout via Git.
```console
$ git fetch upstream '+refs/pull/2562/head:refs/remotes/upstream/pr/2562'
From https://github.com/lima-vm/lima
 * [new ref]           refs/pull/2562/head -> upstream/pr/2562
$ git checkout upstream/pr/2562
HEAD is now at 5223ad9e vz: support kernel image
```

**Comment by norio-nomura** [comment]
I found the official documentation.
https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally?tool=cli

Steps according to the official documentation:
```console
$ git fetch upstream pull/2562/head:vz-support-kernel-image
From https://github.com/lima-vm/lima
 * [new ref]           refs/pull/2562/head -> vz-support-kernel-image
$ git switch vz-support-kernel-image
Switched to branch 'vz-support-kernel-image'
```

**Comment by afbjorklund** [comment]
> > Instead of requiring PRs to be squashed before merging, wouldn't it be better to switch to using [squash merging](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests)?
> 
> It is fine with me, but I don't know if this is acceptable to @AkihiroSuda.

I think it is better if it is up to the author, some commits are relevant while others are not (e.g. merge from master)

**Comment by afbjorklund** [comment]
> @afbjorklund @AkihiroSuda Do you still have any concerns that need to be addressed, or can this be merged now?

I think the result looks more complicated now, but then again I thought that the Kconfig before was "simple" so.... 😀

My previous feedback was just that I was missing simple things like "make exe" or "make test" or coverage etc.
For the goal of making the regular install faster, I just created the `.config` file and kept it in my git workspace.

* https://github.com/lima-vm/lima/issues/2641

As long as there are no regressions, I'm OK with it.

**Comment by norio-nomura** [comment]
squashed and rebased on master.

**Review by jandubois — APPROVED** [review]
Thanks, LGTM

**Comment by norio-nomura** [comment]
Thanks! 🙏🏻

**Pull request merged**